dev_cmds/run-integration-tests.ts

```diff
@@ -1,41 +1,44 @@
-/**
- * This script runs the integration tests for devcmd (and devcmd-cli). Currently,
- * this means mostly: building the package, publishing to a local
- * npm repository, installing devcmd-cli, and installing devcmd in the examples.
- *
- * Some possible next steps:
- * - use a test framework (e.g. jest) to wrap the tests (for reporting etc.)
- * - extract some common functionality
- */
-
-import { execPiped, execToString, runAsyncMain } from "devcmd";
-import { red, green, inverse } from "kleur/colors";
-import fs from "fs-extra";
+import { execPiped, runAsyncMain } from "devcmd";
 import path from "path";
-import { DEVCMD_COMMAND, DOCKER_COMMAND, NPM_COMMAND } from "./utils/commands";
+import fs from "fs-extra";
+import { DEVCMD_COMMAND, DOCKER_COMMAND } from "./utils/commands";
+import { devcmdCliPackageDir, devcmdPackageDir, dockerMountDir, verdaccioConfigDir } from "./utils/paths";
+import { npmPack, NpmPackResult } from "./utils/npm-utils";
 import {
-  devcmdCliPackageDir,
-  devcmdPackageDir,
-  multiplePackageJsonsExampleDir,
-  repoRoot,
-  singlePackageJsonExampleDir,
-} from "./utils/paths";
+  createIntegrationTestGroups,
+  LOCAL_REGISTRY_URL,
+  printTestReport,
+  runIntegrationTests,
+  VERDACCIO_CONTAINER_NAME,
+  VERDACCIO_STORAGE_VOLUME_NAME,
+} from "./integration-tests/integration-test-harness";
+import { delay } from "./utils/delay";
+import { integrationTestGroupFactories } from "./integration-tests/integration-test-group-factories";
+
+async function main(): Promise<void> {
+  await execPiped({ command: DEVCMD_COMMAND, args: ["build-all"] });
+  const packedDevcmdCli = await npmPack(devcmdCliPackageDir);
+  const packedDevcmd = await npmPack(devcmdPackageDir);
 
-const VERDACCIO_CONTAINER_NAME = "devcmd_verdaccio";
-const VERDACCIO_STORAGE_VOLUME_NAME = "devcmd_verdaccio_storage";
-const LOCAL_REGISTRY_URL = "http://0.0.0.0:4873";
+  const { tempImageName } = await createIntegrationTestBaseImage(packedDevcmd, packedDevcmdCli);
 
-const verdaccioConfigDir = path.resolve(repoRoot, "verdaccio");
-const dockerMountDir = path.resolve(repoRoot, "docker-mount");
+  /*
+   * Wondering how to add integration test cases or groups?
+   *
+   * See integration-tests/README.md
+   */
 
-async function main() {
-  await execPiped({ command: DEVCMD_COMMAND, args: ["build-all"] });
+  const testGroups = createIntegrationTestGroups(packedDevcmdCli, integrationTestGroupFactories);
```

Comment:
actually, it would be nicer not to pass `packedDevcmdCli` as a dependency into the factories, but include it as an argument to test execution below instead. that way, tests can be instantiated even without a having packed anything first (e.g. to offer a selection of tests via an interactive menu or filtering by command line args)

Reply:
done